openmc/dagmc.py

```diff
@@ -0,0 +1,617 @@
+        file
+
+        Parameters
+        ----------
+        geom_type : str
+            The type of geometric entity to count. One of {'Volume', 'Surface'}. Returns
+            the runtime number of voumes in the DAGMC model (includes implicit complement).
+
+        Returns
+        -------
+        int
+            Number of geometry elements of the specified type
+        """
+        cv.check_value('geometry type', geom_type, ('volume', 'surface'))
+
+        def decode_str_tag(tag_val):
+            return tag_val.tobytes().decode().replace('\x00', '')
+
+        with h5py.File(self.filename) as dagmc_file:
+            category_data = dagmc_file['tstt/tags/CATEGORY/values']
+            category_strs = map(decode_str_tag, category_data)
+            n = sum([v == geom_type.capitalize() for v in category_strs])
+
+            # check for presence of an implicit complement in the file and
+            # increment the number of cells if it doesn't exist
+            if geom_type == 'volume':
+                name_data = dagmc_file['tstt/tags/NAME/values']
+                name_strs = map(decode_str_tag, name_data)
+                if not sum(['impl_complement' in n for n in name_strs]):
+                    n += 1
+        return n
+
+    @property
+    def n_cells(self):
+        return self._n_geom_elements('volume')
+
+    @property
+    def n_surfaces(self):
+        return self._n_geom_elements('surface')
+
+    def create_xml_subelement(self, xml_element, memo=None):
+        if memo is None:
+            memo = set()
+
+        if self in memo:
+            return
+
+        memo.add(self)
+
+        # Ensure that the material overrides are up-t-date
+        self.build_overide_mat_from_cells()
+
+        # Set xml element values
+        dagmc_element = ET.Element('dagmc_universe')
+        dagmc_element.set('id', str(self.id))
+
+        if self.auto_geom_ids:
+            dagmc_element.set('auto_geom_ids', 'true')
+        if self.auto_mat_ids:
+            dagmc_element.set('auto_mat_ids', 'true')
+        dagmc_element.set('filename', str(self.filename))
+        if self.material_overrides:
+            mat_element = ET.Element('material_overrides')
+            for key in self.material_overrides:
+                print(key, self.material_overrides[key])
+                mat_element.set('id_{}'.format(key), ';'.join(
+                    t for t in self.material_overrides[key]))
+            dagmc_element.append(mat_element)
+        xml_element.append(dagmc_element)
+
+    def build_overide_mat_from_cells(self):
+        """
+        Builds the material override dictionary for cells with multiple instances.
+
+        Returns:
+            None
+        """
+        self.material_overrides = {}
+        for cell in self.cells.values():
+            if isinstance(cell.fill, Iterable):
+                self.material_overrides[str(cell.id)] = [mat.name for mat in cell.fill if mat]
```

Comment:
Here the keys of the `material_overrides` are set to be the strings of cell IDs, but above material names are used. Here's my preferred key-value setup:

keys: Cell IDs (or `openmc.Cell` instances if we want to be fancy, and I think we do)
values: an `Iterable` of `openmc.Material` instances